
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>BPNet Documentatation &#8212; BPNet 0.0.1 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">BPNet 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="bpnet-documentatation">
<h1>BPNet Documentatation<a class="headerlink" href="#bpnet-documentatation" title="Permalink to this headline">¶</a></h1>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p><code class="code bash docutils literal notranslate"><span class="pre">git</span> <span class="pre">clone</span> <span class="pre">https://github.com/kundajelab/basepairmodels.git</span></code>.
<code class="code bash docutils literal notranslate"><span class="name builtin"><span class="pre">cd</span></span> <span class="pre">basepairmodels</span></code>
<code class="code bash docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">.</span></code></p>
</div>
<div class="section" id="api-documentation">
<h2>API Documentation<a class="headerlink" href="#api-documentation" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="command-line-interface">
<h2>Command Line Interface<a class="headerlink" href="#command-line-interface" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
</div>
<div class="section" id="bpnettrainer">
<h3>bpnettrainer<a class="headerlink" href="#bpnettrainer" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">bpnettrainer</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">batch</span><span class="o">-</span><span class="n">size</span> <span class="n">BATCH_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">epochs</span> <span class="n">EPOCHS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">learning</span><span class="o">-</span><span class="n">rate</span> <span class="n">LEARNING_RATE</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">learning</span><span class="o">-</span><span class="n">rate</span> <span class="n">MIN_LEARNING_RATE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">early</span><span class="o">-</span><span class="n">stopping</span><span class="o">-</span><span class="n">patience</span> <span class="n">EARLY_STOPPING_PATIENCE</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">early</span><span class="o">-</span><span class="n">stopping</span><span class="o">-</span><span class="nb">min</span><span class="o">-</span><span class="n">delta</span> <span class="n">EARLY_STOPPING_MIN_DELTA</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">reduce_lr_on_plateau</span><span class="o">-</span><span class="n">patience</span> <span class="n">REDUCE_LR_ON_PLATEAU_PATIENCE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">train</span><span class="o">-</span><span class="n">peaks</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">train</span><span class="o">-</span><span class="n">intervals</span> <span class="n">TRAIN_INTERVALS</span> <span class="p">[</span><span class="n">TRAIN_INTERVALS</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">train</span><span class="o">-</span><span class="n">random</span> <span class="n">TRAIN_RANDOM</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">model</span><span class="o">-</span><span class="n">arch</span><span class="o">-</span><span class="n">name</span> <span class="n">MODEL_ARCH_NAME</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">filters</span> <span class="n">FILTERS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">counts</span><span class="o">-</span><span class="n">loss</span><span class="o">-</span><span class="n">weight</span> <span class="n">COUNTS_LOSS_WEIGHT</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">threads</span> <span class="n">THREADS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">gpus</span> <span class="n">GPUS</span><span class="p">]</span> <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">genome</span> <span class="n">REFERENCE_GENOME</span> <span class="o">--</span><span class="n">chrom</span><span class="o">-</span><span class="n">sizes</span> <span class="n">CHROM_SIZES</span> <span class="p">[</span><span class="o">--</span><span class="n">splits</span> <span class="n">SPLITS</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="n">OUTPUT_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">tag</span><span class="o">-</span><span class="n">length</span> <span class="n">TAG_LENGTH</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">time</span><span class="o">-</span><span class="n">zone</span> <span class="n">TIME_ZONE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">automate</span><span class="o">-</span><span class="n">filenames</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">other</span><span class="o">-</span><span class="n">tags</span> <span class="n">OTHER_TAGS</span> <span class="p">[</span><span class="n">OTHER_TAGS</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">model</span><span class="o">-</span><span class="n">output</span><span class="o">-</span><span class="n">filename</span> <span class="n">MODEL_OUTPUT_FILENAME</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="n">seq</span><span class="o">-</span><span class="nb">len</span> <span class="n">INPUT_SEQ_LEN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">len</span> <span class="n">OUTPUT_LEN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">jitter</span> <span class="n">MAX_JITTER</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">reverse</span><span class="o">-</span><span class="n">complement</span><span class="o">-</span><span class="n">augmentation</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">negative</span><span class="o">-</span><span class="n">sampling</span><span class="o">-</span><span class="n">rate</span> <span class="n">NEGATIVE_SAMPLING_RATE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">dir</span> <span class="n">INPUT_DIR</span><span class="p">]</span>
                    <span class="p">[</span><span class="o">--</span><span class="n">stranded</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">has</span><span class="o">-</span><span class="n">control</span><span class="p">]</span> <span class="o">--</span><span class="n">chroms</span> <span class="n">CHROMS</span> <span class="p">[</span><span class="n">CHROMS</span> <span class="o">...</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">exclude</span><span class="o">-</span><span class="n">chroms</span> <span class="n">EXCLUDE_CHROMS</span> <span class="p">[</span><span class="n">EXCLUDE_CHROMS</span> <span class="o">...</span><span class="p">]]</span>
</pre></div>
</div>
<div class="section" id="Named Arguments">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--batch-size, -b</kbd></dt>
<dd><p>training batch size</p>
<p>Default: 64</p>
</dd>
<dt><kbd>--epochs, -e</kbd></dt>
<dd><p>number of training epochs</p>
<p>Default: 100</p>
</dd>
<dt><kbd>--learning-rate, -L</kbd></dt>
<dd><p>learning rate for Adam optimizer</p>
<p>Default: 0.004</p>
</dd>
<dt><kbd>--min-learning-rate, -l</kbd></dt>
<dd><p>min learning rate for Adam optimizer</p>
<p>Default: 0.0001</p>
</dd>
<dt><kbd>--early-stopping-patience</kbd></dt>
<dd><p>patience value for early stopping callback</p>
<p>Default: 5</p>
</dd>
<dt><kbd>--early-stopping-min-delta</kbd></dt>
<dd><p>minimum change in the validation loss to qualify as an improvement</p>
<p>Default: 0.001</p>
</dd>
<dt><kbd>--reduce_lr_on_plateau-patience</kbd></dt>
<dd><p>patience value for ReduceLROnPlateau callback</p>
<p>Default: 2</p>
</dd>
<dt><kbd>--train-peaks</kbd></dt>
<dd><p>train on peaks contained in the peaks.bed files</p>
<p>Default: False</p>
</dd>
<dt><kbd>--train-intervals</kbd></dt>
<dd><p>train on positions at equal intervals, specify two values &lt;num_positions&gt; &lt;step&gt;. If num_positions is -1 then positions across the entire chromosome are used (superceded by –train-peaks)</p>
<p>Default: [-1, 200]</p>
</dd>
<dt><kbd>--train-random</kbd></dt>
<dd><p>train on randomly sampled positions on the chromosomes. Specify an int value &lt;num_positions&gt;. num_positions has to be a positive number.</p>
<p>Default: 10000</p>
</dd>
<dt><kbd>--model-arch-name</kbd></dt>
<dd><p>the name of the model architesture that will be used in training (the name that will be used to fetch the model from model_archs)</p>
<p>Default: “BPNet”</p>
</dd>
<dt><kbd>--filters, -f</kbd></dt>
<dd><p>number of filters to use in BPNet</p>
<p>Default: 64</p>
</dd>
<dt><kbd>--counts-loss-weight, -w</kbd></dt>
<dd><p>Weight for counts mse loss</p>
<p>Default: 100.0</p>
</dd>
<dt><kbd>--threads, -t</kbd></dt>
<dd><p>number of parallel threads for batch generation</p>
<p>Default: 10</p>
</dd>
<dt><kbd>--gpus, -p</kbd></dt>
<dd><p>number of gpus to use</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--reference-genome, -g</kbd></dt>
<dd><p>number of gpus to use</p>
<p>Default: 1</p>
</dd>
<dt><kbd>--chrom-sizes, -c</kbd></dt>
<dd><p>path to chromosome sizes file</p>
</dd>
<dt><kbd>--splits, -s</kbd></dt>
<dd><p>string name of the function from ‘experiments’ that returns the validation and test splits to use</p>
<p>Default: “10_human_val_test_splits”</p>
</dd>
<dt><kbd>--output-dir, -d</kbd></dt>
<dd><p>destination directory to store the model</p>
<p>Default: “.”</p>
</dd>
<dt><kbd>--tag-length</kbd></dt>
<dd><p>length of the alphanumeric tag for the model file name (applies if –automate-filenames option is used)</p>
<p>Default: 6</p>
</dd>
<dt><kbd>--time-zone</kbd></dt>
<dd><p>time zone to use for timestamping model directories (applies if –automate-filenames option is used)</p>
<p>Default: “US/Pacific”</p>
</dd>
<dt><kbd>--automate-filenames</kbd></dt>
<dd><p>specify if the model output directory and filename should be auto generated</p>
<p>Default: False</p>
</dd>
<dt><kbd>--other-tags</kbd></dt>
<dd><p>list of additional tags to be added as suffix to the filenames</p>
<p>Default: []</p>
</dd>
<dt><kbd>--model-output-filename</kbd></dt>
<dd><p>basename of the model file without the .h5 extension (required if –automate-filenames is not used)</p>
</dd>
<dt><kbd>--input-seq-len</kbd></dt>
<dd><p>length of input DNA sequence</p>
<p>Default: 1346</p>
</dd>
<dt><kbd>--output-len</kbd></dt>
<dd><p>length of output profile</p>
<p>Default: 1000</p>
</dd>
<dt><kbd>--max-jitter</kbd></dt>
<dd><p>maximum value for randomized jitter to offset the peaks from the exact center of the input</p>
<p>Default: 128</p>
</dd>
<dt><kbd>--reverse-complement-augmentation</kbd></dt>
<dd><p>enable reverse complement augmentation</p>
<p>Default: True</p>
</dd>
<dt><kbd>--negative-sampling-rate</kbd></dt>
<dd><p>number of negatives to sample for every positive peak</p>
<p>Default: 0.0</p>
</dd>
<dt><kbd>--input-dir, -i</kbd></dt>
<dd><p>input directory containing bigWigs and peaks</p>
<p>Default: “.”</p>
</dd>
<dt><kbd>--stranded</kbd></dt>
<dd><p>specify if the input data is stranded or unstranded</p>
<p>Default: False</p>
</dd>
<dt><kbd>--has-control</kbd></dt>
<dd><p>specify if the input data has controls</p>
<p>Default: False</p>
</dd>
<dt><kbd>--chroms</kbd></dt>
<dd><p>master list of chromosomes for the genome</p>
</dd>
<dt><kbd>--exclude-chroms</kbd></dt>
<dd><p>list of chromosomes to be excluded</p>
<p>Default: []</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="predict">
<h3>predict<a class="headerlink" href="#predict" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">predict</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">batch</span><span class="o">-</span><span class="n">size</span> <span class="n">BATCH_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="n">seq</span><span class="o">-</span><span class="nb">len</span> <span class="n">INPUT_SEQ_LEN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">len</span> <span class="n">OUTPUT_LEN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">predict</span><span class="o">-</span><span class="n">peaks</span><span class="p">]</span>
               <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">genome</span> <span class="n">REFERENCE_GENOME</span> <span class="o">--</span><span class="n">chrom</span><span class="o">-</span><span class="n">sizes</span> <span class="n">CHROM_SIZES</span> <span class="o">--</span><span class="n">chroms</span> <span class="n">CHROMS</span> <span class="p">[</span><span class="n">CHROMS</span> <span class="o">...</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">DATA_DIR</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">stranded</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">has</span><span class="o">-</span><span class="n">control</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">model</span> <span class="n">MODEL</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">model</span><span class="o">-</span><span class="n">name</span> <span class="n">MODEL_NAME</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">model</span><span class="o">-</span><span class="nb">dir</span> <span class="n">MODEL_DIR</span><span class="p">]</span> <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="n">OUTPUT_DIR</span>
               <span class="p">[</span><span class="o">--</span><span class="n">automate</span><span class="o">-</span><span class="n">filenames</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">time</span><span class="o">-</span><span class="n">zone</span> <span class="n">TIME_ZONE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">exponentiate</span><span class="o">-</span><span class="n">counts</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">window</span><span class="o">-</span><span class="n">size</span> <span class="n">OUTPUT_WINDOW_SIZE</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">other</span><span class="o">-</span><span class="n">tags</span> <span class="n">OTHER_TAGS</span> <span class="p">[</span><span class="n">OTHER_TAGS</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">write</span><span class="o">-</span><span class="n">buffer</span><span class="o">-</span><span class="n">size</span> <span class="n">WRITE_BUFFER_SIZE</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Named Arguments">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--batch-size, -b</kbd></dt>
<dd><p>test batch size</p>
<p>Default: 64</p>
</dd>
<dt><kbd>--input-seq-len</kbd></dt>
<dd><p>length of input DNA sequence</p>
<p>Default: 1346</p>
</dd>
<dt><kbd>--output-len</kbd></dt>
<dd><p>length of output profile</p>
<p>Default: 1000</p>
</dd>
<dt><kbd>--predict-peaks</kbd></dt>
<dd><p>generate predictions only on the peaks contained in the peaks.bed files</p>
<p>Default: False</p>
</dd>
<dt><kbd>--reference-genome, -g</kbd></dt>
<dd><p>the path to the reference genome fasta file</p>
</dd>
<dt><kbd>--chrom-sizes, -s</kbd></dt>
<dd><p>path to chromosome sizes file</p>
</dd>
<dt><kbd>--chroms, -c</kbd></dt>
<dd><p>list of test chromosomes for prediction</p>
</dd>
<dt><kbd>--data-dir</kbd></dt>
<dd><p>input directory containing bigWigs and peaks</p>
<p>Default: “.”</p>
</dd>
<dt><kbd>--stranded</kbd></dt>
<dd><p>specify if the input data is stranded or unstranded (i.e in case –has-control is True)</p>
<p>Default: False</p>
</dd>
<dt><kbd>--has-control</kbd></dt>
<dd><p>specify if the input data has controls</p>
<p>Default: False</p>
</dd>
<dt><kbd>--model, -m</kbd></dt>
<dd><p>path to the .h5 model file</p>
</dd>
<dt><kbd>--model-name</kbd></dt>
<dd><p>the name of the model that will be used in for predictions</p>
<p>Default: “BPNet”</p>
</dd>
<dt><kbd>--model-dir</kbd></dt>
<dd><p>directory where .h5 model files are stored</p>
</dd>
<dt><kbd>--output-dir, -o</kbd></dt>
<dd><p>destination directory to store predictions as a bigWig file</p>
</dd>
<dt><kbd>--automate-filenames</kbd></dt>
<dd><p>specify if the predictions output should be stored in a timestamped subdirectory within –output-dir</p>
<p>Default: False</p>
</dd>
<dt><kbd>--time-zone</kbd></dt>
<dd><p>time zone to use for timestamping model directories</p>
<p>Default: “US/Pacific”</p>
</dd>
<dt><kbd>--exponentiate-counts</kbd></dt>
<dd><p>specify if the predicted counts should be exponentiated before writing to the bigWig files</p>
<p>Default: False</p>
</dd>
<dt><kbd>--output-window-size</kbd></dt>
<dd><p>size of the central window of the output profile predictions that will be written to the bigWig files</p>
<p>Default: 1000</p>
</dd>
<dt><kbd>--other-tags</kbd></dt>
<dd><p>list of additional tags to be added as suffix to the filenames</p>
<p>Default: []</p>
</dd>
<dt><kbd>--write-buffer-size</kbd></dt>
<dd><p>size of the write buffer to store predictions before writing to bigWig files</p>
<p>Default: 10000</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="metrics">
<h3>metrics<a class="headerlink" href="#metrics" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">metrics</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">profileA</span> <span class="n">PROFILEA</span> <span class="o">--</span><span class="n">profileB</span> <span class="n">PROFILEB</span> <span class="p">[</span><span class="o">--</span><span class="n">smooth</span><span class="o">-</span><span class="n">profileA</span> <span class="n">SMOOTH_PROFILEA</span> <span class="p">[</span><span class="n">SMOOTH_PROFILEA</span> <span class="o">...</span><span class="p">]]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">smooth</span><span class="o">-</span><span class="n">profileB</span> <span class="n">SMOOTH_PROFILEB</span> <span class="p">[</span><span class="n">SMOOTH_PROFILEB</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">countsA</span> <span class="n">COUNTSA</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">countsB</span> <span class="n">COUNTSB</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">apply</span><span class="o">-</span><span class="n">softmax</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">profileA</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">apply</span><span class="o">-</span><span class="n">softmax</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">profileB</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">metrics</span><span class="o">-</span><span class="n">seq</span><span class="o">-</span><span class="nb">len</span> <span class="n">METRICS_SEQ_LEN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">peaks</span> <span class="n">PEAKS</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">step</span><span class="o">-</span><span class="n">size</span> <span class="n">STEP_SIZE</span><span class="p">]</span> <span class="o">--</span><span class="n">chroms</span> <span class="n">CHROMS</span> <span class="p">[</span><span class="n">CHROMS</span> <span class="o">...</span><span class="p">]</span> <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="n">OUTPUT_DIR</span> <span class="p">[</span><span class="o">--</span><span class="n">automate</span><span class="o">-</span><span class="n">filenames</span><span class="p">]</span>
               <span class="p">[</span><span class="o">--</span><span class="n">time</span><span class="o">-</span><span class="n">zone</span> <span class="n">TIME_ZONE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">other</span><span class="o">-</span><span class="n">tags</span> <span class="n">OTHER_TAGS</span> <span class="p">[</span><span class="n">OTHER_TAGS</span> <span class="o">...</span><span class="p">]]</span> <span class="o">--</span><span class="n">chrom</span><span class="o">-</span><span class="n">sizes</span> <span class="n">CHROM_SIZES</span>
</pre></div>
</div>
<div class="section" id="Named Arguments">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--profileA, -A</kbd></dt>
<dd><p>the bigWig with ground truth values or a replicate</p>
</dd>
<dt><kbd>--profileB, -B</kbd></dt>
<dd><p>the bigWig with predicted values or the second replicate</p>
</dd>
<dt><kbd>--smooth-profileA</kbd></dt>
<dd><p>a list of two items, sigma and window width for gaussian smoothing of profileA before computing metrics. Empty list indicates nosmoothing</p>
<p>Default: []</p>
</dd>
<dt><kbd>--smooth-profileB</kbd></dt>
<dd><p>a list of two items, sigma and window width for gaussian smoothing of profileB before computing metrics. Empty list indicates nosmoothing</p>
<p>Default: []</p>
</dd>
<dt><kbd>--countsA</kbd></dt>
<dd><p>the bigWig with region counts assigned to each base (the counts track that is produced by the predict script). This is optional.</p>
</dd>
<dt><kbd>--countsB</kbd></dt>
<dd><p>the bigWig with region counts assigned to each base (the counts track that is produced by the predict script). This is optional.</p>
</dd>
<dt><kbd>--apply-softmax-to-profileA</kbd></dt>
<dd><p>apply softmax to profileA before computingmetrics (in casees where profileA is logits)</p>
<p>Default: False</p>
</dd>
<dt><kbd>--apply-softmax-to-profileB</kbd></dt>
<dd><p>apply softmax to profileB before computingmetrics (in casees where profileB is logits)</p>
<p>Default: False</p>
</dd>
<dt><kbd>--metrics-seq-len</kbd></dt>
<dd><p>the length of the sequence over which to compute the metrics</p>
<p>Default: 1000</p>
</dd>
<dt><kbd>--peaks</kbd></dt>
<dd><p>the path to the file containing</p>
</dd>
<dt><kbd>--step-size</kbd></dt>
<dd><p>the step size for genome wide metrics</p>
<p>Default: 50</p>
</dd>
<dt><kbd>--chroms, -c</kbd></dt>
<dd><p>list of test chromosomes to compute metrics</p>
</dd>
<dt><kbd>--output-dir, -o</kbd></dt>
<dd><p>destination directory to store metrics results</p>
</dd>
<dt><kbd>--automate-filenames</kbd></dt>
<dd><p>specify if the metrics output should be stored in a timestamped subdirectory within –output-dir</p>
<p>Default: False</p>
</dd>
<dt><kbd>--time-zone</kbd></dt>
<dd><p>time zone to use for timestamping output directories</p>
<p>Default: “US/Pacific”</p>
</dd>
<dt><kbd>--other-tags</kbd></dt>
<dd><p>list of additional tags to be added as suffix to the filenames</p>
<p>Default: []</p>
</dd>
<dt><kbd>--chrom-sizes, -s</kbd></dt>
<dd><p>path to chromosome sizes file</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="interpret">
<h3>interpret<a class="headerlink" href="#interpret" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">interpret</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="n">seq</span><span class="o">-</span><span class="nb">len</span> <span class="n">INPUT_SEQ_LEN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">control</span><span class="o">-</span><span class="nb">len</span> <span class="n">CONTROL_LEN</span><span class="p">]</span> <span class="o">--</span><span class="n">model</span> <span class="n">MODEL</span> <span class="o">--</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span> <span class="n">BED_FILE</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">presort</span><span class="o">-</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">sample</span> <span class="n">SAMPLE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">chroms</span> <span class="n">CHROMS</span> <span class="p">[</span><span class="n">CHROMS</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">controls</span> <span class="n">CONTROLS</span> <span class="p">[</span><span class="n">CONTROLS</span> <span class="o">...</span><span class="p">]]</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">control</span><span class="o">-</span><span class="n">smoothing</span> <span class="n">CONTROL_SMOOTHING</span> <span class="p">[</span><span class="n">CONTROL_SMOOTHING</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">task</span><span class="o">-</span><span class="nb">id</span> <span class="n">TASK_ID</span><span class="p">]</span> <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="n">OUTPUT_DIR</span>
                 <span class="p">[</span><span class="o">--</span><span class="n">time</span><span class="o">-</span><span class="n">zone</span> <span class="n">TIME_ZONE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">other</span><span class="o">-</span><span class="n">tags</span> <span class="n">OTHER_TAGS</span> <span class="p">[</span><span class="n">OTHER_TAGS</span> <span class="o">...</span><span class="p">]]</span> <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">genome</span> <span class="n">REFERENCE_GENOME</span>
</pre></div>
</div>
<div class="section" id="Named Arguments">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>--input-seq-len, -l</kbd></dt>
<dd><p>the length of the input sequence to the model</p>
<p>Default: 1346</p>
</dd>
<dt><kbd>--control-len</kbd></dt>
<dd><p>the length of the control input to the model</p>
<p>Default: 1000</p>
</dd>
<dt><kbd>--model, -m</kbd></dt>
<dd><p>the path to the model (.h5) file</p>
</dd>
<dt><kbd>--bed-file, -b</kbd></dt>
<dd><p>the path to the bed file containing postions at which the model should be interpreted</p>
</dd>
<dt><kbd>--presort-bed-file</kbd></dt>
<dd><p>specify if the –bed-file should be sorted, in descending order of strongest peaks, before processing</p>
<p>Default: False</p>
</dd>
<dt><kbd>--sample</kbd></dt>
<dd><p>the number of samples to randomly sample from the bed file</p>
</dd>
<dt><kbd>--chroms</kbd></dt>
<dd><p>list of chroms on which the contribution score are to be computed. Empty list implies all chromosomes</p>
<p>Default: []</p>
</dd>
<dt><kbd>--controls, -c</kbd></dt>
<dd><p>list of paths to control bigWig files. If unstranded specify a single file path, if stranded then a plus &amp; minus bigWig file, in that order, must be specified.</p>
</dd>
<dt><kbd>--control-smoothing</kbd></dt>
<dd><p>list of smoothing values for control inputs to the model</p>
<p>Default: [1, 50]</p>
</dd>
<dt><kbd>--task-id</kbd></dt>
<dd><p>In the multitask case the integer sequence number of the task for which the interpretation scores should be computed, for single task use 0.</p>
<p>Default: 0</p>
</dd>
<dt><kbd>--output-dir, -o</kbd></dt>
<dd><p>destination directory to store the interpretation scores</p>
</dd>
<dt><kbd>--time-zone, -t</kbd></dt>
<dd><p>time zone to use for timestamping output directories</p>
<p>Default: “US/Pacific”</p>
</dd>
<dt><kbd>--other-tags</kbd></dt>
<dd><p>list of additional tags to be added as suffix to the filenames</p>
<p>Default: []</p>
</dd>
<dt><kbd>--reference-genome, -g</kbd></dt>
<dd><p>number of gpus to use</p>
<p>Default: 1</p>
</dd>
</dl>
</div>
</div>
<div class="section" id="run-modisco">
<h3>run_modisco<a class="headerlink" href="#run-modisco" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">run_modisco</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span> <span class="n">PATH_TO_IMP_SCORES_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="n">PROFILE_OR_COUNTS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SAVE_PATH</span><span class="p">]</span>
</pre></div>
</div>
<div class="section" id="Named Arguments">
<h4>Named Arguments<a class="headerlink" href="#Named Arguments" title="Permalink to this headline">¶</a></h4>
<dl class="option-list">
<dt><kbd>-d, --path_to_imp_scores_dir</kbd></dt>
<dd><p>Path to the importance scores</p>
</dd>
<dt><kbd>-p, --profile_or_counts</kbd></dt>
<dd><p>scoring method to use, profile or counts scores</p>
</dd>
<dt><kbd>-s, --save_path</kbd></dt>
<dd><p>Enter a save directory. If not provided, it is saved in the logdir</p>
</dd>
</dl>
</div>
</div>
</div>
<div class="section" id="modules">
<h2>Modules<a class="headerlink" href="#modules" title="Permalink to this headline">¶</a></h2>
<div class="section" id="module-basepairmodels.cli.losses">
<span id="loss-functions"></span><h3>Loss Functions<a class="headerlink" href="#module-basepairmodels.cli.losses" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt id="basepairmodels.cli.losses.MultichannelMultinomialNLL">
<em class="property">class </em><code class="sig-prename descclassname">basepairmodels.cli.losses.</code><code class="sig-name descname">MultichannelMultinomialNLL</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.losses.MultichannelMultinomialNLL" title="Permalink to this definition">¶</a></dt>
<dd><p>Class to compute combined loss from ‘n’ tasks</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em>) – the number of channels / tasks</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="basepairmodels.cli.losses.multinomial_nll">
<code class="sig-prename descclassname">basepairmodels.cli.losses.</code><code class="sig-name descname">multinomial_nll</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">true_counts</span></em>, <em class="sig-param"><span class="n">logits</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.losses.multinomial_nll" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the multinomial negative log-likelihood
:param true_counts: observed count values
:param logits: predicted logit values</p>
</dd></dl>

</div>
<div class="section" id="module-basepairmodels.cli.MTBatchGenerator">
<span id="dataloaders"></span><h3>DataLoaders<a class="headerlink" href="#module-basepairmodels.cli.MTBatchGenerator" title="Permalink to this headline">¶</a></h3>
<dl class="py class">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBPNetBatchGenerator">
<em class="property">class </em><code class="sig-prename descclassname">basepairmodels.cli.MTBatchGenerator.</code><code class="sig-name descname">MTBPNetBatchGenerator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tasks</span></em>, <em class="sig-param"><span class="n">reference</span></em>, <em class="sig-param"><span class="n">chrom_sizes</span></em>, <em class="sig-param"><span class="n">chroms</span></em>, <em class="sig-param"><span class="n">batchgen_params</span></em>, <em class="sig-param"><span class="n">name_str</span></em>, <em class="sig-param"><span class="n">stranded</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">num_threads</span><span class="o">=</span><span class="default_value">28</span></em>, <em class="sig-param"><span class="n">epochs</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">batch_size</span><span class="o">=</span><span class="default_value">64</span></em>, <em class="sig-param"><span class="n">shuffle</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'train'</span></em>, <em class="sig-param"><span class="n">control_smoothing</span><span class="o">=</span><span class="default_value">[1, 50]</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBPNetBatchGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>MultiTask batch data generation for BPNet</p>
<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBPNetBatchGenerator.generate_batch">
<code class="sig-name descname">generate_batch</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">coords</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBPNetBatchGenerator.generate_batch" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate one batch of inputs and outputs</p>
<p>For all coordinates in “coords” fetch sequences &amp;
one hot encode the sequences. Fetch corresponding
signal values (for e.g. from a bigwig file),
and aggregate if specified. Package the one hot
encoded sequences and the ouput values as a tuple.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>coords</strong> (<em>pandas.DataFrame</em>) – dataframe with ‘chrom’ and
‘pos’ columns specifying the chromosome and the
coordinate</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A batch tuple with one hot encoded sequences
and corresponding outputs</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py attribute">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBPNetSumAllBatchGenerator">
<code class="sig-prename descclassname">basepairmodels.cli.MTBatchGenerator.</code><code class="sig-name descname">MTBPNetSumAllBatchGenerator</code><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBPNetSumAllBatchGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <a class="reference internal" href="#basepairmodels.cli.MTBatchGenerator.MTBPNetBatchGenerator" title="basepairmodels.cli.MTBatchGenerator.MTBPNetBatchGenerator"><code class="xref py py-class docutils literal notranslate"><span class="pre">basepairmodels.cli.MTBatchGenerator.MTBPNetBatchGenerator</span></code></a></p>
</dd></dl>

<dl class="py class">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator">
<em class="property">class </em><code class="sig-prename descclassname">basepairmodels.cli.MTBatchGenerator.</code><code class="sig-name descname">MTBatchGenerator</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tasks</span></em>, <em class="sig-param"><span class="n">reference</span></em>, <em class="sig-param"><span class="n">chrom_sizes</span></em>, <em class="sig-param"><span class="n">chroms</span></em>, <em class="sig-param"><span class="n">batchgen_params</span></em>, <em class="sig-param"><span class="n">name_str</span></em>, <em class="sig-param"><span class="n">stranded</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">num_threads</span><span class="o">=</span><span class="default_value">28</span></em>, <em class="sig-param"><span class="n">epochs</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">batch_size</span><span class="o">=</span><span class="default_value">64</span></em>, <em class="sig-param"><span class="n">shuffle</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'train'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>MultiTask batch data generation</p>
<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.epoch_run">
<code class="sig-name descname">epoch_run</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.epoch_run" title="Permalink to this definition">¶</a></dt>
<dd><p>Manage batch generation processes &amp; threads
for one epoch</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas.DataFrame</em>) – dataframe with ‘chrom’ &amp;
‘pos’ columns</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>—</p>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.gen">
<code class="sig-name descname">gen</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.gen" title="Permalink to this definition">¶</a></dt>
<dd><p>generator function to yield batches of data</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>pandas.DataFrame</em>) – dataframe with ‘chrom’ &amp;
‘pos’ columns</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>—</p>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.generate_batch">
<code class="sig-name descname">generate_batch</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">coords</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.generate_batch" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate one batch of inputs and outputs</p>
</dd></dl>

<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.get_negative_batch">
<code class="sig-name descname">get_negative_batch</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.get_negative_batch" title="Permalink to this definition">¶</a></dt>
<dd><p>get chrom positions for the negative samples using
uniform random sampling from across the all chromosomes
in self.chroms</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><dl class="simple">
<dt>dataframe of coordinates</dt><dd><p>(‘chrom’ &amp; ‘pos’)</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.proc_target">
<code class="sig-name descname">proc_target</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">coords_df</span></em>, <em class="sig-param"><span class="n">mpq</span></em>, <em class="sig-param"><span class="n">proc_idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.proc_target" title="Permalink to this definition">¶</a></dt>
<dd><p>Function that will be executed in a separate process.
Takes a dataframe of peak coordinates and parses them in
batches, to get one hot encoded sequences and corresponding
outputs, and adds the batches to the multiprocessing queue.
Optionally, samples negative locations and adds them to
each batch</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>coords_df</strong> (<em>pandas.DataFrame</em>) – dataframe containing
the chrom &amp; peak pos</p></li>
<li><p><strong>mpq</strong> (<em>multiprocessing.Queue</em>) – The multiprocessing queue
to hold the batches</p></li>
</ul>
</dd>
</dl>
<dl class="simple">
<dt>Return</dt><dd><p>—</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.set_early_stopping">
<code class="sig-name descname">set_early_stopping</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.set_early_stopping" title="Permalink to this definition">¶</a></dt>
<dd><p>Set early stopping flag to True</p>
<p>Args: None</p>
<p>Returns: None</p>
</dd></dl>

<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.set_ready_for_next_epoch">
<code class="sig-name descname">set_ready_for_next_epoch</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">val</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.set_ready_for_next_epoch" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the variable that controls batch generation
for the next epoch</p>
<blockquote>
<div><p>Args: None</p>
<p>Returns: None</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.set_stop">
<code class="sig-name descname">set_stop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.set_stop" title="Permalink to this definition">¶</a></dt>
<dd><p>Set stop Flag to True</p>
<p>Args: None</p>
<p>Returns: None</p>
</dd></dl>

<dl class="py method">
<dt id="basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.stealer">
<code class="sig-name descname">stealer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mpq</span></em>, <em class="sig-param"><span class="n">q</span></em>, <em class="sig-param"><span class="n">num_batches</span></em>, <em class="sig-param"><span class="n">thread_id</span></em><span class="sig-paren">)</span><a class="headerlink" href="#basepairmodels.cli.MTBatchGenerator.MTBatchGenerator.stealer" title="Permalink to this definition">¶</a></dt>
<dd><p>Thread target function to “get” (steal) from the
multiprocessing queue and “put” in the regular queue</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mpq</strong> (<em>multiprocessing.Queue</em>) – The multiprocessing queue
to steal from</p></li>
<li><p><strong>q</strong> (<em>Queue</em>) – The regular queue to put the batch into</p></li>
<li><p><strong>num_batches</strong> (<em>int</em>) – the number of batches to “steal”
from the mp queue</p></li>
<li><p><strong>thread_id</strong> (<em>int</em>) – thread id for debugging purposes</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>—</p>
</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="indices-and-tables">
<h3>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">BPNet Documentatation</a><ul>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#api-documentation">API Documentation</a></li>
<li><a class="reference internal" href="#command-line-interface">Command Line Interface</a><ul>
<li><a class="reference internal" href="#bpnettrainer">bpnettrainer</a><ul>
<li><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#predict">predict</a><ul>
<li><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#metrics">metrics</a><ul>
<li><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#interpret">interpret</a><ul>
<li><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
<li><a class="reference internal" href="#run-modisco">run_modisco</a><ul>
<li><a class="reference internal" href="#Named Arguments">Named Arguments</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#modules">Modules</a><ul>
<li><a class="reference internal" href="#module-basepairmodels.cli.losses">Loss Functions</a></li>
<li><a class="reference internal" href="#module-basepairmodels.cli.MTBatchGenerator">DataLoaders</a></li>
<li><a class="reference internal" href="#indices-and-tables">Indices and tables</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/index.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="#">BPNet 0.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Zahoor Zafrulla, Ziga Avsec, Avanti Shrikumar, Anshul Kundaje.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.0.3.
    </div>
  </body>
</html>